File: src/Talabi.Core/DTOs/ApiResponse.cs
namespace Talabi.Core.DTOs;

public class ApiResponse<T>
{
    public bool Success { get; set; }
    public string? Message { get; set; }
    public T? Data { get; set; }
    public string? ErrorCode { get; set; }
    public List<string>? Errors { get; set; }

    public ApiResponse() { }

    public ApiResponse(T data, string? message = null)
    {
        Success = true;
        Data = data;
        Message = message;
    }

    public ApiResponse(string message, string? errorCode = null, List<string>? errors = null)
    {
        Success = false;
        Message = message;
        ErrorCode = errorCode;
        Errors = errors;
    }
}

File: mobile/lib/models/api_response.dart
class ApiResponse<T> {
  ApiResponse({
    required this.success,
    this.message,
    this.data,
    this.errorCode,
    this.errors,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? json) fromJsonT,
  ) {
    return ApiResponse(
      success: json['success'] ?? false,
      message: json['message'],
      data: json['data'] != null ? fromJsonT(json['data']) : null,
      errorCode: json['errorCode'],
      errors: json['errors'] != null ? List<String>.from(json['errors']) : null,
    );
  }
  final bool success;
  final String? message;
  final T? data;
  final String? errorCode;
  final List<String>? errors;
}
